### PR TITLE
integration: add a test to verify creating cluster on hub

### PR DESCRIPTION
**Pull request by chanwit**
This is a test for `swarm create` and `swarm c`. These command should return a proper cluster token, hexadecimal string of size 32.

Signed-off-by: Chanwit Kaewkasi <chanwit@gmail.com>

### CODE REVIEW

**Comment by bacongobbler** [comment]
I'm liking this new integration test library :+1:

**Comment by chanwit** [comment]
Me too! it's a cool stuff done by @aluzzardi

**Comment by aluzzardi** [comment]
I believe once we get #582 merged we could update this to do a full test of the token discovery.

**Comment by chanwit** [comment]
Relates to #644

**Comment by chanwit** [comment]
@aluzzardi I think I need your advise on how should we continue the `token` tests?

**Comment by aluzzardi** [comment]
@vieux: Would there be a way to destroy clusters? Worried that this test will create thousands of clusters overtime.

**Comment by abronan** [comment]
@aluzzardi There is a `DELETE` method on the Hub discovery service to destroy clusters. We can use it to delete the created cluster at the end of the test (or in `teardown()`).

**Comment by abronan** [comment]
Well for now you have to use:

```
curl -X DELETE https://discovery-stage.hub.docker.com/v1/clusters/<cluster-id>
```

As there is no `swarm destroy` command..

**Comment by chanwit** [comment]
@aluzzardi could you please check this command:

`godep go run main.go create | echo $?`
On my machine, it returns `2`.

While `$GOBIN/swarm create | echo $?` returns `0`.

With this, we cannot properly check `$status` via `godep go run`. Any idea?

**Comment by chanwit** [comment]
@aluzzardi never mind my ping. I think it's caused by the Hub does not response sometime. `godep go run main.go create` now returns `0`.

File: test/integration/create-cluster-on-hub.bats [hunk 1]
```diff
@@ -0,0 +1,29 @@
+#!/usr/bin/env bats
+
+load helpers
+
+function token_cleanup() {
+	curl -X DELETE https://discovery-stage.hub.docker.com/v1/clusters/$1
```
Comment (abronan) [inline test/integration/create-cluster-on-hub.bats]: We should probably wait for a `swarm destroy` command so you can actually test the creation and then the destruction of a cluster.

Comment (chanwit) [inline test/integration/create-cluster-on-hub.bats]: @abronan OK. I'll wait for it.

Comment (aluzzardi) [inline test/integration/create-cluster-on-hub.bats]: I'm not sure if we're going to get the `swarm destroy` soon enough and we need this integration test.

For me it's an LGTM

@abronan @vieux Thoughts?

Comment (abronan) [inline test/integration/create-cluster-on-hub.bats]: @aluzzardi Well, we can change it if we introduce something like a `swarm destroy` command in the future, so this is not an issue. Let's merge it if we need it now :+1:

**Comment by chanwit** [comment]
Please do not merge yet. I still can't do 100% regression for all tests.

**Comment by chanwit** [comment]
done. could you please review again.

File: test/integration/helpers.bash [hunk 2]
```diff
@@ -88,7 +88,7 @@ function swarm_join() {
 		SWARM_JOIN_PID[$i]=$!
 		((++i))
 	done
-	wait_until_swarm_joined $i
+	wait_until_swarm_joined $i "$@"
```
Comment (aluzzardi) [inline test/integration/helpers.bash]: I'm wondering if `wait_until_swarm_joined` could use the new `retry` function (just like `wait_until_reachable`)

Maybe something like ```retry 10 1 [ `swarm list "$@" | wc -l` -eq $i ]```

File: test/integration/create-cluster-on-hub.bats [hunk 3]
```diff
@@ -0,0 +1,29 @@
+#!/usr/bin/env bats
```
Comment (aluzzardi) [inline test/integration/create-cluster-on-hub.bats]: Also, regarding the name, I'd call it `token-discovery.bats` to be consistent with the other integration tests you wrote (`file-discovery.bats`, etc)

File: test/integration/helpers.bash (hunk 2 above)
Comment (chanwit) [inline test/integration/helpers.bash]: @aluzzardi I haven't looked at your new `retry` yet. Will be trying it out.

File: test/integration/create-cluster-on-hub.bats (hunk 3 above)
Comment (chanwit) [inline test/integration/create-cluster-on-hub.bats]: sure :)

File: test/integration/helpers.bash [hunk 4]
```diff
@@ -88,24 +88,7 @@ function swarm_join() {
...
-	until [ $attempts -ge $max_attempts ]; do
-		run docker -H $SWARM_HOST info
-		if [[ "${lines[3]}" == *"Nodes: $1"* ]]; then
-			break
-		fi 
-		echo "Checking if joined successfully for the $((++attempts)) time" >&2
-		sleep 1
-	done
-	[[ $attempts -lt $max_attempts ]]
+	retry 30 1 [ `swarm list "$@" | wc -l` -eq $i ]
```
Comment (chanwit) [inline test/integration/helpers.bash]: @aluzzardi I set this large to let the test pass as my machine is quite far from Hub. Maybe `30` would be reduced when you've got integration tests run on CI.

File: test/integration/token-discovery.bats [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+	start_docker 2
+
+	TOKEN=$(swarm create)
+	[ "$status" -eq 0 ]
+	[[ ${TOKEN} =~ ^[0-9a-f]{32}$ ]]
+
+	swarm_manage token://$TOKEN
+	swarm_join   token://$TOKEN
+
+	retry 10 1 docker_swarm info && [[ "${lines[3]}" == *"Nodes: 2"* ]]
```
Comment (chanwit) [inline test/integration/token-discovery.bats]: @aluzzardi This too to avoid network time-out problem. Not sure if you're OK with this.

Comment (aluzzardi) [inline test/integration/token-discovery.bats]: Oh, I see the problem - basically `docker list` returns 2 items faster than what swarm sees, right?

In that case, what about changing the test in swarm_join from:
`[ `swarm list "$@" | wc -l` -eq $i ]`
to something that runs docker_swarm info. Maybe:
[ -n `docker_swarm info | grep "Nodes: $i" ]

File: test/integration/helpers.bash [hunk 6]
```diff
@@ -88,24 +88,7 @@ function swarm_join() {
...
-	until [ $attempts -ge $max_attempts ]; do
-		run docker -H $SWARM_HOST info
-		if [[ "${lines[3]}" == *"Nodes: $1"* ]]; then
-			break
-		fi 
-		echo "Checking if joined successfully for the $((++attempts)) time" >&2
-		sleep 1
-	done
-	[[ $attempts -lt $max_attempts ]]
+	retry 30 1 [ `docker_swarm info | grep "Nodes: $i"` -eq 0 ]
```
Comment (chanwit) [inline test/integration/helpers.bash]: @aluzzardi should this do?

Comment (aluzzardi) [inline test/integration/helpers.bash]: Yup, I think so! Not sure what `-eq 0` is doing though? `[ -n ``docker_swarm ... ``]` should work too I guess (`-n` = the length of STRING is nonzero)

Comment (chanwit) [inline test/integration/helpers.bash]: it's testing the exit code of the expression in backticks.
`if [[ `echo 2 | grep -q 2` -eq 0 ]]; then echo yes; else echo no; fi` shows yes and 
`if [[ `echo 2 | grep -q 2` -eq 1 ]]; then echo yes; else echo no; fi` shows no for me.

Maybe I should add `-q` to it.

**Comment by chanwit** [comment]
regression goes well. PTAL.
@aluzzardi @vieux @abronan

**Comment by aluzzardi** [comment]
Awesome!

As you may have noticed, we run those automatically now (`janky` in the list of GH checks)

LGTM

**Comment by chanwit** [comment]
Wow, that `janky` is super awesome! I've check its console output and all discovery tests are doing well there. Super!

**Comment by aluzzardi** [comment]
@chanwit yep! If any integration test failed, this PR would be marked as failed (just like when travis fails because of broken tests)

**Comment by aluzzardi** [comment]
Related to #465

**Comment by chanwit** [comment]
@aluzzardi I think you just did a circular reference here.

**Comment by aluzzardi** [comment]
Woops.

#644

**Pull request merged**